feat: introduce flag to set the http response timeout value

DennisDenuto opened the pull request:
- increased the default timeout from 10s -> 30s
- refactor: Remove copied golang http.DefaultTransport
- It is easier to see what has been 'changed' from the DefaultTransport.

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
Refactor: Remove copied golang http.DefaultTransport

- It is easier to see what has been 'changed' from the DefaultTransport.

Authored-by: Dennis Leon <leonde@vmware.com>

cppforlife commented on pkg/imgpkg/registry/registry.go:
```diff
@@ -222,25 +221,13 @@ func newHTTPTransport(opts Opts) (*http.Transport, error) {
-			InsecureSkipVerify: (opts.VerifyCerts == false),
-		},
-	}, nil
+	clonedDefaultTransport := http.DefaultTransport.(*http.Transport).Clone()
```
the only downside i see here is that if something changes default transport this code will be affected. do we know if anything changing default transport anywhere?

DennisDenuto replied:
the upside is that as we upgrade the http package, we get any changes / updates made to the DefaultTransport. (we avoid drifting from DefaultTransport)

i.e. since inlining the DefaultTransport, we drifted in 2 fields:
- `ForceAttemptHTTP2` setting to true 
- `DialContext.DualStack` has been deprecated 

sounds like a trade-off to me. While the risk you outline is true, currently nothing is mutating the `DefaultTransport` in our codebase or any of our libraries. 

My assumption is that the chance of drifting from stdlib is higher than the chance something will mutate DefaultTransport, and that's why I think this tradeoff is 'worth it'. 

Another assumption is that we don't want to drift from stdlib because the go folk are making reasonable sane changes to the DefaultTransport that we *want* in imgpkg.

DennisDenuto committed:
feat: add --registry-response-header-timeout flag

- used to specify how long requests to the registry should wait for a
http response (headers) from the registry.
- increased the default timeout from 10s -> 30s

Authored-by: Dennis Leon <leonde@vmware.com>

nithinmy commented:
Not sure if I was doing it right, but I just tried to push a local branch and got this error -
```
08:38:00.829: [carvel-imgpkg] git -c credential.helper= -c core.quotepath=false -c log.showSignature=false push --progress --porcelain origin refs/heads/config-cli-responseheadertimeout:config-cli-responseheadertimeout --set-upstream
remote: Permission to vmware-tanzu/carvel-imgpkg.git denied to nithinmy.
fatal: unable to access 'https://github.com/vmware-tanzu/carvel-imgpkg.git/': The requested URL returned error: 403
```

cppforlife commented on pkg/imgpkg/cmd/registry_flags.go:
```diff
@@ -19,6 +20,8 @@ type RegistryFlags struct {
 	Token    string
 	Anon     bool
+
+	ResponseTimeout time.Duration
```
this should be ResponseHeaderTimeout. ResponseTimeout implies that it's a timeout for the entire response which isnt the case of how we are using this value.

DennisDenuto replied:
ok done

cppforlife commented on pkg/imgpkg/cmd/registry_flags.go:
```diff
@@ -30,6 +33,8 @@ func (r *RegistryFlags) Set(cmd *cobra.Command) {
 	cmd.Flags().StringVar(&r.Token, "registry-token", "", "Set token for auth ($IMGPKG_TOKEN)")
 	cmd.Flags().BoolVar(&r.Anon, "registry-anon", false, "Set anonymous auth ($IMGPKG_ANON)")
+
+	cmd.Flags().DurationVar(&r.ResponseTimeout, "registry-response-timeout", 30*time.Second, "Maximum time to allow a request to wait for a http response from the registry (ms|s|m|h)")
```
registry-response-timeout -> registry-response-header-timeout

DennisDenuto replied:
ok done
